File: app/src/main/java/com/smoothradio/radio/core/ui/RadioViewModel.java
package com.smoothradio.radio.core.ui;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.smoothradio.radio.core.data.repository.PlayerPreferencesRepository;
import com.smoothradio.radio.core.data.repository.RadioLinkRepository;
import com.smoothradio.radio.core.data.repository.RadioRepository;
import com.smoothradio.radio.core.util.Resource;
import com.smoothradio.radio.core.model.RadioStation;
import com.smoothradio.radio.service.StreamService;

import java.util.List;

public class RadioViewModel extends AndroidViewModel {
    private final RadioLinkRepository radioLinkRepository;
    private final PlayerPreferencesRepository prefsRepo;
    private final RadioRepository repository;

    private LiveData<Resource<List<String>>> localRadioLinksLiveData;
    private LiveData<Resource<Integer>> stationIdLiveData;
    private LiveData<List<RadioStation>> allStations;
    private final LiveData<List<RadioStation>> favoriteStations;

    private final MutableLiveData<RadioStation> selectedStation = new MutableLiveData<>();
    private final MutableLiveData<Integer> currentPage = new MutableLiveData<>(0);
    private final MutableLiveData<Boolean> updateMiniPlayerEvent = new MutableLiveData<>();


    public RadioViewModel(@NonNull Application application) {
        super(application);
        radioLinkRepository = new RadioLinkRepository();
        prefsRepo = new PlayerPreferencesRepository(application.getApplicationContext());
        repository = new RadioRepository(application);

        allStations = repository.getAllStations();
        favoriteStations = repository.getFavoriteStations();
    }

    // -------------------
    // Radio Station Logic
    // -------------------
    public LiveData<List<RadioStation>> getAllStations() {
        return allStations;
    }
    public LiveData<List<RadioStation>> getFavoriteStations() {
        return favoriteStations;
    }
    public void savePlayingStationId(int id) {
        repository.setPlayingStation(id);
    }
    public LiveData<RadioStation> getPlayingStation() {
        return repository.getPlayingStation();
    }
    public void insertStations(List<RadioStation> stations) {
        repository.insertStations(stations);
    }

    public void updateFavoriteStatus(int id, boolean isFav) {
        repository.updateFavoriteStatus(id, isFav);
    }
    // -----------------------
    // Radio Links Logic
    // -----------------------

    // Fetch remote links from Firestore and update file only if needed
    public void getRemoteLinks() {
        localRadioLinksLiveData = radioLinkRepository.getRemoteStreamLinks();
    }

    public LiveData<Resource<List<String>>> getRemoteLinksLiveData() {
        return localRadioLinksLiveData;
    }

    public void removeStreamLinkListener() {
        radioLinkRepository.removeListener();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        radioLinkRepository.removeListener();
    }
    // -----------------------
    // Preferences Logic
    // -----------------------
    public LiveData<Resource<Integer>> getStationIdLivedata() {
        return stationIdLiveData;
    }

    public void getStationId() {
        stationIdLiveData = prefsRepo.getIdLiveData();
    }

    public void saveStationId(int id) {
        prefsRepo.setId(id);
    }
    public void reloadStationId() {
        prefsRepo.reloadStationId();
    }


    // -----------------------
    // PLayer Logic
    // -----------------------
    public void setSelectedStation(RadioStation station) {
        selectedStation.setValue(station);
    }

    public LiveData<RadioStation> getSelectedStation() {
        return selectedStation;
    }

    //-------------------------
    // ViewPager Current Page Logic
    //-------------------------
    public LiveData<Integer> getCurrentPage() {
        return currentPage;
    }

    public void setCurrentPage(int page) {
        currentPage.setValue(page);
    }

    //------------------------
    // Update Mini Player Trigger
    //------------------------
    public LiveData<Boolean> getOnRemoteLinksLoadedEvent() {
        return updateMiniPlayerEvent;
    }

    // Call this when remote links are loaded successfully
    public void onRemoteLinksLoaded() {
        updateMiniPlayerEvent.setValue(true);
    }
}

File: app/src/main/java/com/smoothradio/radio/core/ui/RadioViewModel.kt
package com.smoothradio.radio.core.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.smoothradio.radio.core.data.repository.RadioLinkRepository
import com.smoothradio.radio.core.data.repository.RadioRepository
import com.smoothradio.radio.core.model.RadioStation
import com.smoothradio.radio.core.util.Resource
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.shareIn
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class RadioViewModel @Inject constructor(
    application: Application,
    private val radioLinkRepository: RadioLinkRepository,
    private val radioRepository: RadioRepository
) : AndroidViewModel(application) {

    val allStations = radioRepository.allStations

    val playingStation = radioRepository.playingStation


    val remoteLinks: Flow<Resource<List<String>>> = radioLinkRepository
        .getRemoteStreamLinksFlow()
        .shareIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
        )

    val favoriteStations = radioRepository.favoriteStations

    private val _selectedStation = MutableSharedFlow<RadioStation?>(replay = 0)
    val selectedStation: SharedFlow<RadioStation?> = _selectedStation

    private val _currentPage = MutableStateFlow(0)
    val currentPage: StateFlow<Int> = _currentPage.asStateFlow()

    // Radio Station Logic
    fun savePlayingStationId(id: Int) {
        viewModelScope.launch {
            radioRepository.setPlayingStation(id)
        }
    }

    fun insertStations(stations: List<RadioStation>) {
        viewModelScope.launch {
            radioRepository.insertStations(stations)
        }
    }

    fun updateFavoriteStatus(id: Int, isFav: Boolean) {
        viewModelScope.launch {
            radioRepository.updateFavoriteStatus(id, isFav)
        }
    }
    // Player Logic
    fun setSelectedStation(station: RadioStation?) {
        viewModelScope.launch {
            _selectedStation.emit(station)
        }
    }
    // ViewPager Current Page Logic
    fun setCurrentPage(page: Int) {
        _currentPage.value = page
    }

    override fun onCleared() {
        super.onCleared()
        radioLinkRepository.clear()
    }
}